packages/workspace/src/browser/workspace-commands.ts

```diff
@@ -194,45 +209,93 @@ export class WorkspaceCommandContribution implements CommandContribution {
     }
 }
 
-export interface UriCommandHandler {
-    execute(uri: URI, ...args: any[]): any;
+export interface UriCommandHandler<T extends URI | URI[]> {
+
+    // tslint:disable-next-line:no-any
+    execute(uri: T, ...args: any[]): any;
+
+    // tslint:disable-next-line:no-any
     isEnabled?(uri: URI, ...args: any[]): boolean;
+
+    // tslint:disable-next-line:no-any
     isVisible?(uri: URI, ...args: any[]): boolean;
+
+}
+
+export interface SingleUriCommandHandler extends UriCommandHandler<URI> {
+
 }
-export class FileSystemCommandHandler implements CommandHandler {
+
+export interface MultiUriCommandHandler extends UriCommandHandler<URI[]> {
+
+}
+
+export class UriAwareCommandHandler<T extends URI | URI[]> implements CommandHandler {
+
     constructor(
         protected readonly selectionService: SelectionService,
-        protected readonly handler: UriCommandHandler
+        protected readonly handler: UriCommandHandler<T>,
+        protected readonly multi: boolean = false
     ) { }
 
-    protected getUri(...args: any[]): URI | undefined {
+    // tslint:disable-next-line:no-any
+    protected getUri(...args: any[]): T | undefined {
         if (args && args[0] instanceof URI) {
-            return args[0];
+            return this.multi ? [args[0]] : args[0];
+        }
+        const { selection } = this.selectionService;
+        if (UriSelection.is(selection)) {
+            return (this.multi ? [selection.uri] : selection.uri) as T;
         }
-        return UriSelection.getUri(this.selectionService.selection);
+        if (StructuredSelection.is(selection)) {
```

Comment:
What is the reason to introduce `StructuredSelection`? with arrays you can do the same:
```ts
if (Array.is(selection)) {
    if (this.multi) {
        for(const item of selection.filter(UriSelection.is)) {
            uris.push(item.uri)            
        }
    }
    if (UriSelection.is(selection[0]) {
        return selection[0].uri as T;
    }
}
```

Reply:
> StructuredSelection

I prefer having a dedicated type for that.

Reply:
but why? it makes all clients verbose